storm-multilang/dotnet/src/main/dotnet/Dotnet.Storm.Adapter/Components/BaseBolt.cs

```diff
@@ -0,0 +1,133 @@
+﻿using Dotnet.Storm.Adapter.Extensions;
+using Dotnet.Storm.Adapter.Messaging;
+using System;
+using System.Collections.Generic;
+
+namespace Dotnet.Storm.Adapter.Components
+{
+    public abstract class BaseBolt : Component
+    {
+        protected new class LocalStorm : Component.LocalStorm
+        {
+            public void Ack(string id)
+            {
+                GlobalStorm.Send(new AckMessage(id));
+            }
+
+            public void Fail(string id)
+            {
+                GlobalStorm.Send(new FailMessage(id));
+            }
+
+            public void Emit(List<object> tuple, string stream = "default", long task = 0, List<string> anchors = null, bool needTaskIds = false)
+            {
+                VerificationResult result = VerifyOutput(stream, tuple);
+
+                if (result.IsError)
+                {
+                    Logger.Error($"Validation failed: {result} in {tuple}.");
+                }
+                else
+                {
+                    BoltTuple message = new BoltTuple()
+                    {
+                        Task = task,
+                        Stream = stream,
+                        Tuple = tuple,
+                        Anchors = anchors,
+                        NeedTaskIds = needTaskIds
+                    };
+
+                    GlobalStorm.Send(message);
+                }
+            }
+        }
+
+        private bool running = true;
+
+        public BaseBolt()
+        {
+            Storm = new LocalStorm();
+        }
+
+        internal override void Start()
+        {
+            while(running)
+            {
+                Message message = GlobalStorm.Receive<ExecuteTuple>();
+                if (message != null)
+                {
+                    // there are only two options: task_ids and tuple to execute
+                    if (message is TaskIdsMessage ids)
+                    {
+                        OnTaskIds(this, new TaskIds(ids));
+                    }
+                    if (message is ExecuteTuple tuple)
+                    {
+                        VerificationResult result = Storm.VerifyInput(tuple.Component, tuple.Stream, tuple.Tuple);
+
+                        if (result.IsError)
+                        {
+                            Logger.Error($"Validation failed: {result} in {tuple}.");
+                        }
+                        else
+                        {
+                            switch (tuple.Stream)
+                            {
+                                case "__heartbeat":
+                                    DoHeartbeat();
+                                    break;
+                                case "__tick":
+                                    DoTick();
```

Comment:
Shouldn't check of tick messages be moved to Execute()? Otherwise tick message is only checked only when a component starts?

Reply:
That's not true.  The Start method is big infinite loop. Tick tuple is specific. It is not required to be processed. It is introduced for some specific scenarios. For example you have a bold which should persist it's state periodically. The tick tuple is kind of signal to bolt to do such periodic tasks. that's why it is an event.